Fix buffer leaks in MP codec and in tests

pderop opened the pull request:
Motivation:

So far, the _LeakDetector_ was not used when running the tests (with the -P leak profile), because _-Dio.netty5.buffer.leakDetectionEnabled=true_ was missing in the _test.argLine_ property of the leak profile (in the parent pom).

Now, using the option, many leaks are reported. This PR fixes the remaining leaks and will unblock the other #14 PR which is currently failing because of the buffer leaks.

These leaks only concerns this MP codec, not the old MP from the old Netty 4.x, because in old version, the MP was using some unpooled buffers, and in 4.x, the not-released unpooled buffers are not logged by the LeakDetector.
Now, we are only used pooled buffers, which should all be properly closed.

Motifications:

In the source codec:

- fixed AbstractMemoryHttpData.setContentInternal which now verifies if the current resource content must be closed.  Before, the MemoryAttribute.prepareSend method was calling setContentInternal and then previous content resource was not closed
- fixed HttpPostMultipartRequestDecoder class: The attributes from the currentFieldAttributes list are now closed before emptying the currentFieldAttributes list.
- fixed HttpPostRequestEncoder class: empty chunk buffers are now closed
- fixed HttpPostStandardRequestDecoder: setFinalBuffer is now closing the current attribute buffer (if the attribute is not a memory based (because for memory http data, the getBuffer() method returns a singleton, and does not allocate a copy of the current memory http data buffer)

in the test classes:

fixed many tests where buffers and codec objects were not closed properly.

Result:

The CI should now pass when the -P leak profile is used, and the other #14  PR will be unblocked by this one (the other #14 PR will have to be rebased on top of main, once this PR will be merged).

pderop committed:
Fix buffer leaks in codec and tests

pderop commented:
it would be great if someone could have a look before I can merge this ?

pderop commented:
@chrisvest, @normanmaurer ,

Maybe anyone of you could review this PR before I merge it ?
Only 4 classes from src/main/java are involved by few modifications while all the other less important changes are done in tests. I'll then be able to rebase the other #14, and after all that, this project will be in a releasable state.

thanks !

chrisvest commented on codec-multipart/src/main/java/io/netty/contrib/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:
```diff
@@ -992,11 +1005,35 @@ public void removeHttpDataFromClean(InterfaceHttpData data) {
+    }
+
+    private void cleanMixedAttributes(AsciiString attributeKey) {
+        currentFieldAttributes.computeIfPresent(attributeKey, (charSequence, attribute) -> {
```
`Map.remove(key)` returns the previously associated value, if any. Seems a bit simpler.

pderop replied:
thanks for the suggestion, I just removed the  cleanMixedAttributes(AsciiString attributeKey) method and I used try-with-resource in the cleanMixedAttributes() method in order to close old removed resources, if any. There is now less code.

chrisvest commented on codec-multipart/src/test/java/io/netty/contrib/handler/codec/http/multipart/AbstractTest.java:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 20202 The Netty Project
```
```suggestion
 * Copyright 2022 The Netty Project
```

pderop replied:
thank you, I have corrected the typo.

chrisvest commented on codec-multipart/src/test/java/io/netty/contrib/handler/codec/http/multipart/AbstractTest.java:
```diff
@@ -0,0 +1,34 @@
+    @AfterAll
+    static void afterAll() {
+        if (Boolean.getBoolean("io.netty5.buffer.leakDetectionEnabled")) {
+            IntStream.range(0, 10).forEach(i -> System.gc());
```
10 times sounds a little excessive? The GC itself is (sometimes, depending on the GC) an asynchronous event that is only scheduled by this call.

pderop replied:
Indeed, it's not necessary, just one single GC is enough, I have corrected that.

chrisvest commented on codec-multipart/src/main/java/io/netty/contrib/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java:
```diff
@@ -539,11 +539,16 @@ private void parseBodyAttributes() {
         addHttpData(currentAttribute);
+        if (! currentAttribute.isInMemory() && currentBuf != null && currentBuf.isAccessible()) {
+            // disked based attribute always returned an allocated buffer from getBuffer method, so we need to close
+            currentBuf.close();
```
You may want to investigate if this difference between disk and memory attributes can be eliminated, because that's going to be confusing to people.

pderop replied:
Indeed, the difference of behaviour between disked based and memory based attributes may be confusing; ok I will try to investigate how to improve this.

chrisvest reviewed (commented):
Had some comments on a couple of things that stood out.

pderop committed:
Simplified cleanMixedAttributes

pderop committed:
Fixed typo in copyright year, and only perform one GC, which is sufficient

pderop commented:
Rebased in order to pick up #16

chrisvest commented on benchmarks/run-asyncprof.sh:
```diff
@@ -0,0 +1,18 @@
+CLASS=${2}
+ASYNCPROFILER=/Users/pderop/Work/projects/async-profiler
+java -jar target/microbenchmarks.jar -f 1 -i 2 -wi 2 -r 2 -w 2 -prof async:libPath=$ASYNCPROFILER/build/libasyncProfiler.so\;event=${EVENT} ${CLASS}
+
```
Did you intend to include these scripts?

pderop replied:
oops ! no, this is a mistake, I will remove them in few moment !

chrisvest commented on codec-multipart/src/test/java/io/netty/contrib/handler/codec/http/multipart/AbstractTest.java:
```diff
@@ -0,0 +1,32 @@
+    static void afterAll() {
+        if (Boolean.getBoolean("io.netty5.buffer.leakDetectionEnabled")) {
+            System.gc();
+        }
```
You could also do this as a JUnit 5 extension, and have it be automatically registered: https://junit.org/junit5/docs/current/user-guide/#extensions-registration-automatic

That way you don't force all tests to inherit from an `AbstractTest`.

pderop replied:
ok, I will take a look, thanks

pderop committed:
Added new HttpData.mapBuffer method, fixed two other buffer leaks.

- Introduced a new mapBuffer method in HttpData interface in order to simplify the HttpPostStandardRequestDecoder.setFinalBuffer() method.
- Fixed buffer leak in DiskFileUpload.copy method, and added test in DiskFileUploadTest.testCopy()
- Fixed buffer leak in DiskAttribute.copy method, and added test in HttpDataTest.testCopy()
- Clarified javadoc for HttpData content(), getBuffer() and getChunk(int) methods.

pderop commented:
@chrisvest ,
please wait before reviewing the last commit, indeed, I just saw your feedback, and  I need to remove some scripts which were added by mistake, and I need to also address your feedback regarding junit ... 
thanks !

pderop committed:
Removed scripts which were added by mistake

pderop committed:
Use Junit5 extension in order to run GC once all tests are executed

pderop commented:
So, Junit5 extension is now used, and tests are not extending anymore the _AbstractTest_, which has been renamed to _GCExtension_.

Also, in the 15715ec commit, I did an attempt to simplify the _HttpPostStandardRequestDecoder.setFinalBuffer_ using a new _mapBuffer_ method added in the _HttpData_ interface.

The _mapBuffer_ method allows to return the result of applying any functions to the _HttpData_ content, and takes care of closing the buffer passed to the function if the _HttpData_ is disk based.

However, we still have something which may be cleaned: the buffer ownership of the buffers returned by the _HttpData_ _getBuffer/content_ methods depends on the nature of the _HttpData_:

- for disk based, a copy is returned by the _getBuffer()/content()_ methods; so the user should close the returned buffer.
- but for memory based, the internal buffer is returned by the _getBuffer()/content()_ methods; so if the user closes the returned buffer, the internal buffer will be inaccessible.

For the moment, I'm not sure how to simplify this, so maybe we can consider doing that later if needed, using another PR.

PS: I clarified the javadoc for getBuffer/content/getChunk(int) methods in HttpData interface.

can you re-take a look ?

pderop commented:
PS: in 15715ec, I also fixed two remaining buffer leaks in DiskFileUpload.copy() and iDiskAttribute.copy() methods.

pderop committed:
Reverted the mapBuffer added in previous commits

pderop committed:
Fixed some remaining leaks in tests

pderop commented:
@chrisvest ,

Hello Chris,

I'm back on this old PR. So, I applied your suggestion regarding Junit, I will merge this PR since you gave me a +1.

Please notice that I have reverted the _mapBuffer_ methods which was an attempt to eliminate the differences between disk based and memory based attributes. In fact, I have another solution in mind and it's better to create a **separate** PR for that.

thanks for the reviews.
